[doc] publish the api policy (#47020)
ray-project/ray: doc/source/ray-contribute/api-policy.rst

.. _api-policy:

API Policy
=============

Ray APIs refer to classes, class methods, or functions.
When something is declared as an API, we promise our
users that they can use these APIs to develop their
applications without worrying about changes to these
interfaces between different Ray releases. Declaring
or deprecating an API has a significant impact on the
community.  This document proposes simple policies to
hold Ray contributors accountable to these promises
and manage our user expectations.

For API exposure level, please refer to :ref:`API Stability <api-stability>`.

API Documentation Policy
~~~~~~~~~~~~~~~~~~~~~~~~
Documentation is one of the main channels through which
we expose our APIs to users. If we provide incorrect
information, it can significantly impact the reliability
and maintainability of our users' applications. Based on
the API exposure level, here is the policy to ensure the
accuracy of our information.

.. list-table:: API Documentation Policy
    :widths: 20 16 16 16 16 16
    :header-rows: 1

    * - Policy/Exposure Level
      - Stable Public API
      - Beta Public API
      - Alpha Public API
      - Deprecated API
      - Developer API

>     * - Must this API be documented?

Reviewer: Is it called an annotation or a decorator?

      - Yes
      - Yes
      - Yes
      - Yes
      - Up to the developers
    * - Must a method API be annotated?
      - Yes
      - Yes
      - Yes
      - Yes
      - No annotations mean it is a developer API by default

>     * - Must this API have an end-of-life date declared?

Reviewer: How do you declare and EOL date?

      - No
      - No
      - No
      - Yes. The end-of-life date is a future Ray release version.
      - No

>     * - Can this API be private (either inside the _internal module or has an underscore prefix)?

Reviewer: ```suggestion
    * - Can this API be private, either be inside the `_internal` module, or has an underscore prefix in the name?
```

      - No
      - No
      - No
      - No
      - No

> API Lifecycle Policy

Reviewer: ```suggestion
API lifecycle policy
```

~~~~~~~~~~~~~~~~~~~~
Users have high expectations for certain exposure levels,

> so we need to be cautious when moving APIs between different

Reviewer: ```suggestion
so use cautioun when moving APIs between different
```

> levels. Here is the policy for managing the API exposure lifecycle.

Reviewer: ```suggestion
levels. Below is the policy for managing the API exposure lifecycle.
```

.. list-table:: API Lifecycle Policy
    :widths: 20 16 16 16 16 16
    :header-rows: 1

    * - Policy/Exposure Level
      - Stable Public API
      - Beta Public API
      - Alpha Public API
      - Deprecated API
      - Developer API

>     * - Can this API be promoted to a higher level without any warnings, heads up to users?

Reviewer: ```suggestion
    * - Can you promote this API to a higher level without any warnings or communication to users?
```

      - Yes
      - Yes
      - Yes
      - No
      - Yes

>     * - Can this API be demoted to a lower level? If so then how?

Reviewer: ```suggestion
    * - Can you demote this API to a lower level?
```

>       - Can be demoted to Deprecated only. The API should emit warning messages and a deadline for deprecations in **6 months (or +25 ray minor versions)**.

Reviewer: ```suggestion
      - You can demote it to `Deprecated` only. The API should emit warning messages and you must set a deadline for deprecation in **6 months, or +25 Ray minor versions**.
```

Reviewer: Add an inline link to what a Ray minor version is?

>       - Can be demoted to Deprecated only. The API should emit warning messages and a deadline for deprecations in **3 months (or +12 ray minor versions)**.

Reviewer: ```suggestion
      - You can demote it to `Deprecated` only. The API should emit warning messages and you must set a deadline for deprecations in **3 months or +12 Ray minor versions**.
```

>       - Breaking changes must be both allowed and expected in alpha components, and users must have no expectation of stability.

Reviewer: ```suggestion
      - Users must allow for and expect breaking changes in alpha components, and have no expectations of stability.
```

      - Yes

>       - No annotations mean it is a developer API by default

Reviewer: ```suggestion
      - No. The absence of annotations implies the Developer API level by default.
```

>     * - Can this API remove or change it parameters?

Reviewer: ```suggestion
    * - Can you remove or change this API's parameters?
```

>       - The API should emit warning messages and a deadline for the change in **6 months (or +25 ray minor versions)**. During the transition period, both the new and old parameters are supported

Reviewer: ```suggestion
      - Yes. The API should emit warning messages and you must set a deadline for the end-of-life of the original version that is **6 months or +25 Ray minor versions**. During the transition period, you must support both the new and old parameters.
```

>       - The API should emit warning messages and a deadline for the change in **3 months (or +12 ray minor versions)**. During the transition period, both the new and old parameters are supported

Reviewer: ```suggestion
      - The API should emit warning messages and you must set a deadline for the change in **3 months or +12 Ray minor versions**. During the transition period, you must support both the new and old parameters.
```

>       - Breaking changes must be both allowed and expected in alpha components, and users must have no expectation of stability.

Reviewer: ```suggestion
      - Users must allow for and expect breaking changes in alpha components, and must have no expectations of stability.
```
